### ISSUE TITLE
Status access violation

### ISSUE THREAD

**Issue by leonardodepaula** [parallaxsecond/rust-cryptoki#240]
When I updated from version 0.7.0 to 0.8.0 I ran into this error: (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)

```
fn main() {
    let pkcs11client = Pkcs11::new("C:/Windows/System32/eTPKCS11.dll").unwrap();
    pkcs11client.initialize(CInitializeArgs::OsThreads).unwrap();

    let slot = pkcs11client.get_slots_with_token().unwrap().remove(0);

    let session = pkcs11client.open_ro_session(slot).unwrap();
    session.login(UserType::User, None).unwrap();

    println!("{:?}", session);
}
```

Platform: Windows 11

The code above runs perfectly on version 0.7.0, but it doesn't on version 0.8.0.

**Comment by hug-dev** [comment]
Hey!

I tried your example but it fails with SoftHSM on both versions :( Couldn't find a commit which changed parts that would cause this to fail between 0.7.0 and 0.8.0. If you have time, you could `git bisect` to check...

Which line is failing in your code?
Have you also checked that giving `None` for pin would require `CKF_PROTECTED_AUTHENTICATION_PATH` (as shown [here](https://docs.rs/cryptoki/latest/cryptoki/session/struct.Session.html#method.login))?

**Comment by leonardodepaula** [comment]
Hello, @hug-dev!

First of all, thanks for the reply and the amazing work no the crate.

I'm prototyping here, so git bisect won't help.

This is the line failing:

`
let pkcs11client = Pkcs11::new("C:/Windows/System32/eTPKCS11.dll").unwrap();
`

I also tried setting the pin and the error persisted.

I made some tests and this is the last commit fully working:

`cryptoki = { git = "https://github.com/parallaxsecond/rust-cryptoki.git", rev="024976fc892b96ad72fbc4af38cf9449c42034c4" }`

So, the probable bug was introduced here: https://github.com/parallaxsecond/rust-cryptoki/commit/3ec788c23a49e189d56e7b841d99dc21519b92b1

**Comment by hug-dev** [comment]
ah nice find! Interesting, I did not spot anything in the commit that could have caused this isse :thinking: 
Maybe not the issue but https://github.com/parallaxsecond/rust-cryptoki/pull/230 also changed those parts recently

**Comment by hug-dev** [comment]
Do you also know if your PKCS11 implementation (`eTPKCS11.dll`) implements the PKCS11 3.0 headers? That could be the issue with the commit you linked

**Comment by leonardodepaula** [comment]
> Do you also know if your PKCS11 implementation (`eTPKCS11.dll`) implements the PKCS11 3.0 headers? That could be the issue with the commit you linked (although it was supposed to be retrocompatible)

I don't know. It is provided by Gemalto's Safenet Authentication Client. How can I check that?

I forked this repository and made two tests. Upgraded the pkcs11-headers to the version 3.1 and downgraded the headers to the version 2.40. On both scenarios the error persisted. I don't know the ins and outs of the cryptoki or the concrete implementation of this crate, but I think it indicates that the incompatibility doesn't come from the "cryptoki-sys/vendor/pkcs11.h" file.

**Comment by hug-dev** [comment]
> How can I check that?

At least [here](https://cpl.thalesgroup.com/access-management/security-applications/authentication-client-token-management) in Supported APIs they say PKCS11 v2.20 :cry: 

But I don't see how you would get an error from it..

**Comment by leonardodepaula** [comment]
> I tried your example but it fails with SoftHSM on both versions :( Couldn't find a commit which changed parts that would cause this to fail between 0.7.0 and 0.8.0. If you have time, you could `git bisect` to check...

Are you getting an error with SoftHSM as well? What is the error?

**Comment by hug-dev** [comment]
I get:

```
thread 'main' panicked at cryptoki/examples/generate_key_pair.rs:27:41:
called `Result::unwrap()` on an `Err` value: Pkcs11(ArgumentsBad, Login)
```
on line
```
    session.login(UserType::User, None).unwrap();
```

**Comment by jrozner** [comment]
I'm also running into this on windows 11 using ykcs11

**Comment by hug-dev** [comment]
> I'm also running into this on windows 11 using ykcs11. It works fine using the library on linux and macos.

Do you also have the same offending [commit](https://github.com/parallaxsecond/rust-cryptoki/commit/3ec788c23a49e189d56e7b841d99dc21519b92b1)? I don't have access to a Windows machine to test sadly :cry: 
We should probably run the CI on Windows too

**Comment by jrozner** [comment]
I confirmed that it doesn't work in 3ec788c but does work in it's parent commit 024976f.

**Comment by jrozner** [comment]
I enabled JIT debugging on windows and was able to catch the crash. The crash is here. I'm not clear exactly what this is doing.

![image](https://github.com/user-attachments/assets/50d31396-5a55-4c9b-9262-363ca738c27d)

The root cause is an indirect call to unmapped memory. This is the disassembly.

![image](https://github.com/user-attachments/assets/158728fb-d223-4932-9abc-8c4c7600021e)

rax has 7FFF184196E00000 which is just wrong. The address looks shifted 16bits to the left.

Looking at the memory map, this aligns. Shifting it back would put it in ykcs11 (the pkcs11 library I'm using)
![image](https://github.com/user-attachments/assets/f203d2d6-82dc-43e3-b4d3-5c8ea3392aab)

If I do the shift and look up the address it corresponds to `C_Finalize` within the library. Is that what it's supposed to be calling?
![image](https://github.com/user-attachments/assets/3538841c-ba74-42db-8842-453032bc9fe6)

**Comment by leonardodepaula** [comment]
Great work, @jrozner! I'm travelling with my family and will be able to help in two weeks. Thanks!

**Comment by leonardodepaula** [comment]
@Direktor799, since you commited https://github.com/parallaxsecond/rust-cryptoki/commit/3ec788c23a49e189d56e7b841d99dc21519b92b1, do you have any idea what kind of tests we could run to find what was the breaking change?

**Comment by Direktor799** [comment]
> [@Direktor799](https://github.com/Direktor799), since you commited [3ec788c](https://github.com/parallaxsecond/rust-cryptoki/commit/3ec788c23a49e189d56e7b841d99dc21519b92b1), do you have any idea what kind of tests we could run to find what was the breaking change?

Hi, sorry for the delay, I might have some idea about this. Afaik pkcs11 standard does not define the alignment of the fields inside a struct, but normally vendors tend to use no alignment on Windows platforms at all.

See this from the old v2.4 header.
https://github.com/parallaxsecond/rust-cryptoki/blob/024976fc892b96ad72fbc4af38cf9449c42034c4/cryptoki-sys/pkcs11.h#L79-L82
This does not exist in the new header and adding it back should fix this issue. Does anyone want to try it? Or I can do it in a few weeks.

**Comment by jrozner** [comment]
I put up a PR that makes this change [here](https://github.com/parallaxsecond/rust-cryptoki/pull/245). When I ran the script to re-generate the bindings, all the PTR types are missing. Is there something special I have to do to have bindgen create those? @hug-dev or someone else who is familiar with the build process?

**Comment by hug-dev** [comment]
First time I see something like that! Having looked a bit more at your change, it seems to be only the `_PTR` types declared with the `ULONGDEF` macro which are not created, but the `_PTR`-less types are 🤔 Something wonky with the pre-processor that `bindgen` uses? Maybe the syntax of the macro is not fully compliant?
You could try to first execute a pre-processor on the file (if it works) and then uses that.

About structure packing in general, we got [this](https://github.com/parallaxsecond/rust-cryptoki/issues/70) a while back.

**Comment by gshaurya18** [comment]
> so probably something with macro preprocessing

That seems likely since `ULONGDEF` is a function like macro: https://github.com/parallaxsecond/rust-cryptoki/blob/4cb53df08a4ca0b6d0c1fd8cf368b32377d4c902/cryptoki-sys/vendor/pkcs11.h#L35-L37

and so is `STRUCTDEF`: https://github.com/parallaxsecond/rust-cryptoki/blob/4cb53df08a4ca0b6d0c1fd8cf368b32377d4c902/cryptoki-sys/vendor/pkcs11.h#L1066-L1070

I think adding [`clang_macro_fallback`](https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#method.clang_macro_fallback) to the [builder](https://github.com/parallaxsecond/rust-cryptoki/blob/main/cryptoki-sys/build.rs#L89) will likely get the the missing bindings as well.

Reference bindgen's [Workaround for expansion of function-like macros](https://github.com/rust-lang/rust-bindgen/pull/2779)

**Comment by hug-dev** [comment]
Nice find! I think it's fine to bump `bindgen` and fix that!

**Comment by jrozner** [comment]
I updated bindgen to 0.70.1 and enabled `clang_macro_fallback` in the build.rs. Re-created the bindings and everything seems to be working. I was able to successfully test this with Yubico's ykcs11 library. Changes are in #245 and ready for review. Once this merges, is it possible to get a release?

**Comment by hug-dev** [comment]
Yes we can make a new release afterwards! There has been quite a few improvements!

**Comment by leonardodepaula** [comment]
I just tested the changes in #245 proposed by @jrozner and everything worked just fine for the Safenet Authentication Client (SAC). Thank everyone for the effort here, looking forward for the new release.

I make myself available to run tests on Windows and the SAC stack whenever needed.

**Comment by hug-dev** [comment]
`cryptoki` 0.9.0, which contains this fix, is now available! Thank you all for the help 🥳
